### ISSUE TITLE
Support of graph databases for EF Core

### ISSUE THREAD

**Issue by ohault** [dotnet/efcore#33795]
**Note**: SQL Server graph support is tracked by #8527.

For AI, support of graph databases through in EF Core will become incredibly important.  

I can see it side by side with MongoDB support via MongoDB Query Language (MQL).

Can you please set this request as a high priority and strategic feature on your roadmap for next year release ?

**Comment by wstaelens** [comment]
Duplicate: https://github.com/dotnet/efcore/issues/30180

**Comment by ohault** [comment]
> Duplicate of: #30180
This is not a duplicate. This ticket is not indented to be specific or limited to SQL Server.

**Comment by roji** [comment]
@ohault at this stage, we have no knowledge of what "support of graph databases" would mean independent of any specific database; the graph functionality in the different databases (e.g. SQL Server, MongoDB) seems different enough that it's far from clear what a "cross-database graph feature" or infrastructure would look like. This is why we generally track this sort of capability on a per-database basis.

In addition, note that the MongoDB EF provider is maintained by Mongo themselves, and not in this repo.

Finally, I'd be interested to know what concrete AI-related functionality relates to graph database support - more context from you  (e.g. references) could help prioritize work in this area.

**Comment by wstaelens** [comment]
I think that by having graph database support in EF Core for SQL Server will make small graph databases more accessible in small scale projects without having to setup any cosmos db and having to pay for azure.
A simple sql express might do the trick and would work like a charm. Currently there is no good "free" alternative.

**Comment by wstaelens** [comment]
> I think that by having graph database support in EF Core for SQL Server will make small graph databases more accessible in small scale projects without having to setup any cosmos db and having to pay for azure. A simple sql express might do the trick and would work like a charm. Currently there is no good "free" alternative.


update: I think that sometimes for small projects where you don't know if they might succeed or not (or for local development) you might just want to set up a VM + sql express. or create an additional database in sql express and play/work with graphs.

EF + .net + sql express might be very useful for this.

In case Cosmos db could be self hosted (a bit like cosmos db emulator + gremlin via EF) that might be an alternative way of working.

Many people trust SQL Server and EF because it works and is also rock solid for years. If Graph capabilities are included in both, you get a cheap alternative instead of CosmosDB. In case you reach limits or need more power you will go automatically to CosmosDB, see it as a (Microsoft) opportunity. It would allow existing projects to easily add small graphics by setting up a different EF Graph context and 'upgrade'/'uplift' a lot of projects.

**Comment by roji** [comment]
> Many people trust SQL Server and EF because it works and is also rock solid for years. If Graph capabilities are included in both, you get a cheap alternative instead of CosmosDB.

I'm assuming by CosmosDB you're referring to [Cosmos for Germlin](https://learn.microsoft.com/en-us/azure/cosmos-db/gremlin/introduction). In any case, given that EF doesn't support either SQL Server graph capabilities or Germlin, I'm not sure how anything here is an alternative to anything else. If you want to use either, you can only do that without EF.

**Comment by merddyin** [comment]
The thing that kind of boggles me with regards to the ongoing debate on this is that the use cases and scenarios related to graph and AI should be more than obvious by this point. A key element of many AI solutions is the development of knowledge graphs, which are functionally what is provided by the graph capability within SQL server, as well as other graph DBs. Being able to interact with knowledge graphs, from code, without having to resort to raw DB queries, or migrating to something like Neo4j, would make things far easier.

Outside of the general use case, the specific elements around SQL Server Graph support seems like a no-brainer. The whole concept of the relationships between entities is one that is built into EF, so the only real difference here is that the DB provides a handy mechanism for tracking these relationships, in the form of the Edge tables. The only element that seems to be at all potentially tricky to deal with would be the graph specific query language extensions. The feature has been present 8 years across multiple versions of SQL, and the popularity of AI ensures that it isn't going anywhere. By contrast, SQL Server support for Vector types has barely shipped and there is already support in EFCore for it, theoretically because of the increased interest in AI. Well, graph functionality is also relevant to AI in the form of knowledge graphs, as mentioned above, and it really is something that should have been added long since.

**Comment by wstaelens** [comment]
important is to have this in all versions e.g. SQL Express so that we can set up these small knowledge graphs locally hosted and when it becomes too large, migrate to Azure or Cosmos/Gremlin.

**Comment by roji** [comment]
@merddyin and others, I'm not expert in graph databases and their specific usage in AI. However, while the usage of vector search in AI scenarios is extremely mainstream, widespread and well-understood, I personally have not seen the same level of popularity of using graph search in AI scenarios. That's not to say that the approaches and patterns don't exist (e.g. GraphRAG) - just that they don't seem to be anywhere near as popular/relevant as vector-based search.

One specifically important aspect is that graph data - like traditional relational/SQL data - is fundamentally structured, whereas vector search operates on totally unstructured data. So if your organization happens to already have a graph database set up and you wish to be able to add that as an additional query source for AI (typically alongside vector search) - that makes a lot of sense. But if such a graph database doesn't already exist, the task of creating it is not trivial. I'm aware of approaches to use an LLM to populate your graph database, e.g. [neo4j's LLM knowledge graph builder](https://neo4j.com/developer/genai-ecosystem/llm-graph-builder/)), but that's yet another non-trivial piece of complexity in the data pipeline and I'm not sure how proven it is as an approach.

To summarize, EF doesn't yet support graph support (in SQL Server or otherwise) since we haven't seen the demand for it. I currently don't see a specific reason graph would be important for AI more than other database types (e.g. if you happen to have a redis key-value store, you might also want to query as part of your AI application - just like if you happen to have a graph database). However, I'd be very happy to learn something new here (as I said, I'm no expert).

Finally, don't forget that not everything has to happen via EF APIs - you can use SQL today to perform graph search on SQL Server on any other database and integrate the results with other searches, etc.

**Comment by keithnicholson** [comment]
Maybe there doesn't seem to be a demand, since it doesn't seem like MS wants to support SQL Graph at all. Is there even a product manager for it any more?

What a waste, it held so much promise.

**Comment by roji** [comment]
I'll go ahead and close this as a duplicate of #8527.

**Comment by ohault** [comment]
No to be closed, general graph support and support of graph in SQL Server are not the same. This issue is far wider and important.

**Comment by keithnicholson** [comment]
Agreed!  Please do not close this.  If MS is not going to support SQL Graph as a product, my team needs to move to a 3rd party graph database.  Hence, I still need to interface with it and I would prefer to do it with EF Core.

**Comment by roji** [comment]
I'll reopen, though note that it matters little whether we have one issue or two (for SQL Server vs. in general).

In general, EF's querying story is built around LINQ, which is a great fit for SQL (and also for most document databases), but is probably not a good way to express graph queries (e.g. via Cypher, which seems to be the de-facto graph querying language. I haven't looked too deeply into this, but it'd be worth understanding why exactly people are looking for EF support for graph databases here - not everything related to data makes sense with EF.

**Comment by wstaelens** [comment]
> I'll reopen, though note that it matters little whether we have one issue or two (for SQL Server vs. in general).
> 
> In general, EF's querying story is built around LINQ, which is a great fit for SQL (and also for most document databases), but is probably not a good way to express graph queries (e.g. via Cypher, which seems to be the de-facto graph querying language. I haven't looked too deeply into this, but it'd be worth understanding why exactly people are looking for EF support for graph databases here - not everything related to data makes sense with EF.

Cypher, I thought Gremlin was the de-facto graph querying language.

**Comment by roji** [comment]
> Cypher, I thought Gremlin was the de-facto graph querying language.

I'm really no expert in graph databases or languages - but I do know that Cypher is used by Neo4j, FactorDB, and Apache Age. There's also GQL was seems to be an effort for an official standard, but isn't yet actually supported AFAIK.

Regardless of the specific language... For anyone who wants "EF support for graph databases", it would be great if you could go a bit further and provide an example of what that would actually mean. In other words, how would you expect to query your graph database - would LINQ be involved, how, etc.

**Comment by keithnicholson** [comment]
How would want query?

Yes, LINQ would be used. And that’s the reason for wanting this.  I’m just getting into graphs, as is my team.  We have been doing EF for a long time with Fluent.

That is what we want to do is to is use a set of key extension functions to query without writing dynamic SQL.   I can imagine Matches and Shortest Paths being added. Taking advantage of Last and other key words. I would love to see the structure doing the paths from nodes via connection. The if we need to get data from a relational table, a simple includes.

Add the ability to do mutations, priceless.




Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by roji** [comment]
> Yes, LINQ would be used. And that’s the reason for wanting this.

It's easy enough to say "use LINQ to translate queries to graph database", but my point was **how** such translations would actually look like. LINQ was designed for traditional, SQL-like querying, which is quite different from graph querying, and so may not be the right tool for this.

For example, graph searching supports [bidirectional traversals](https://memgraph.com/docs/querying/read-and-modify-data#bidirectional-traversals):

```cypher
MATCH  (p1:Person)-[r:FRIENDS_WITH]-(p2:Person {name:'Alison'})
RETURN p1, r, p2;
```

This query returns pairs of people, each including Alison and another Person where there's a friendship relationship **in either direction**. Filtering in this way in LINQ would look something like the following:

```c#
_ = context.People
    .Where(p =>
        // This returns people who have an Alison as a friend
        p.Friends.Any(f => f.Name == "Alison")
        // This returns people who some Alison has as a friend
        || context.People.Where(p => p.Name == "Alison").SelectMany(p => p.Friends).Contains(p)))
```

(note that the above only filters, and does not actually project out the Alisons with the friends - that would add more complexity).

This particular example shows that while Cypher has an elegant mechanism for expressing bidirectional (or direction-less) relationships, LINQ is always directional, requiring a complicated duplication. Personally, if you ask me whether I want to write such a complicated LINQ query or the Cypher query just above, I'd choose the Cypher one every day, even if it's a string (and therefore weakly-typed) querying mechanism (via some custom operator). Note also that this is just one feature of Cypher that's problematic to translate (bidirectional relationships).

Now, I haven't spent any actual time thinking about all this - the above is just the first thing that occurred to me. It's possible that if we sit down and think about it, we'd find a LINQ-based mechanism that would both be elegant/terse as the Cypher query it translates into, and offers strong typing in .NET. That's why I'm asking for a little more than "please make LINQ work for graph databases".

But if that's not the case, and the source language (.NET/LINQ) turns out to be considerably more complicated than the target language we're translating into (e.g. Cypher) - as it seems in my above example - then doing such a translation doesn't seem like a good idea.

**Comment by merddyin** [comment]
Personally, I am less worried about the ability to query using LINQ (though that would be nice), and more interested in the ability to create tables with the 'AS NODE' and 'AS EDGE' without needing to resort to raw SQL. In my case, I have a use case where I will need to leverage both regular relational tables for things like application users and roles, as well as graph tables to keep track of complex and changing relationships between entities. Being able to support these with a consistent set of framework and database connectivity options using EFCore would be ideal, but I can't create NODE or EDGE tables, nor can I effectively include them in my definitions of the data model, which means I almost need two separate apps to do one thing.

In terms of the supportability, I seem to recall that there was an individual who actually solved this problem years ago now. He submitted a pull request that was never merged because the architecture got changed. His methodology however, should still provide a viable path to resolve this, and he created said solution in a couple days at most if I recall. Further, from what I have read of the EFCore documentation, keeping track of the relationships between various inter-related tables is something that EFCore already does, even when there are multiple levels of relationship between tables, so this would functionally seem like an evolution of that capability.

I am not suggesting that support for ALL graph databases be added, though it would certainly be nice to have a single ORM for any type of graph or relational database. What I am suggesting here is that a feature that has been part of SQL server since 2012 be supported. As for comments about differentiating code between relational databases, you guys have technically already set a precedent for that when you added support for the not even GA yet vectors feature that is _not_ present for all relational databases supported by EFCore.

**Comment by roji** [comment]
> Personally, I am less worried about the ability to query using LINQ (though that would be nice), and more interested in the ability to create tables with the 'AS NODE' and 'AS EDGE' without needing to resort to raw SQL.

This is something that we repeat quite often: specifically for migration features (i.e. table creation), it's simply unfeasible for EF to support all CREATE TABLE DDL possibilities, across all databases; if you look at SQL Server's CREATE TABLE, you'll see a great deal of possibilities, and that's without even taking other databases into account. EF cannot model all of these for every scenario - that's simply not possible. Just as importantly, we absolutely do recommend that users edit their migrations when needed, and use SQL directly to express whatever they need. One frequent problem is that EF users expect to be able to do absolutely everything - 100% - via EF APIs, and consider any sort of SQL usage to represent a gap or failing of EF; that's simply not the way we see EF.

So while I understand the desire to have EF handle absolutely everything for you, if what you're looking for is simply to create tables with NODE/EDGE, that's something you can already do today with SQL migrations.

**Comment by keithnicholson** [comment]
From the way I am reading the responses, they seem to come across without any empathy.  Why do you think people use EF if not to make coding and databases easier?  Heck that is the promise of every tool provided by Microsoft.

It doesn’t sound like the requests are that extreme as one poster states it seems like it was nearly completed previously with little effort.  There are foundations to do this.

I don’t believe anyone has requested it do it all, but is it unreasonable to consider the most current market driven and those created under the same roof of Microsoft?  It seems more like the team has to solve for every database before it would consider it.  But that is not true with vectors, temporal, and other features available in the company’s own database.

I haven’t been unable to respond to the syntax question yet and I intend to reply.   Yet I would have thought a planning meeting with the PM of SQL Graph would have gone a lot further than anything we could suggest.

If the conveyors to simply kick the idea down the street until demand is so high, EF faces the chicken and the egg question, which came first.
________________________________

**Comment by roji** [comment]
> From the way I am reading the responses, they seem to come across without any empathy.  Why do you think people use EF if not to make coding and databases easier?  Heck that is the promise of every tool provided by Microsoft.

Apologies if my comment came across as lacking empathy - that wasn't my intention. Specifically for migration features, we get a great deal of requests for various DDL support; I'd love to be able to implement each and every one and have EF cover the entire range of all DDL possibilities across all possibilities - but that's simply not possible. More importantly, this happens to be one area where there's a relatively reasonable workaround - edit your migrations and use SQL for that particular feature which EF is missing (this isn't always possible - but it is here). I'm mainly trying to say that in a realistic world of constrained resources, we have to choose where we focus our efforts, and we tend to do that in highly-requested, high-impact areas where no reasonable workaround exists.

Once again, a main point here is that like many other software components, EF will generally not be able to do 100% of what you need from it - the important thing is that you're able to not be blocked by it. An occasional point of frustration for me is when users argue very strongly against **any** usage of SQL in an EF project, sometimes saying that if EF can't do some relatively niche thing that they need, it's effectively completely useless and has no value (that's probably the source of what came across as a lack of empathy - and I apologize).

Note that this isn't unique to migrations: EF also cannot translate all possible LINQ queries - you will occasionally run up against a complicated query form that isn't supported (or is translated in an unsatisfactory way), and your only choice at that point is to drop down to SQL. I'm trying to say that while not ideal, that's OK - a project that has 98% of its queries covered by EF's LINQ provider is already getting great value from EF, even if 2% must be hand-coded in SQL.

> It doesn’t sound like the requests are that extreme as one poster states it seems like it was nearly completed previously with little effort.  There are foundations to do this.

Can you point to this please?

> If the conveyors to simply kick the idea down the street until demand is so high, EF faces the chicken and the egg question, which came first.

I don't think that's true - we generally get good, simple signal from user votes on our issues, without chicken and egg problems. In fact, the SQL Server graph issue (##8527) has garnered 131 votes, which definitely isn't nothing. Unfortunately, at this point it's also relatively few given the probable complexity and implementation time of this feature, and other competing features (but that may change).

**Comment by merddyin** [comment]
As mentioned above, I personally would be more than happy with even the ability to simply create the tables with ASNODE or ASEDGE that I can define in my code. My scenario calls for a mix of relational tables for more static items and graph tables for tracking complex relationships in a dynamic manner, something relational tables just aren't good at. 

Don't get me wrong, as I would totally love to be able to run my queries via LINQ, but I get how that could be a bigger deal, since there is a lot more to work out. I'm mostly trying to avoid things being...messy, when it comes to creating the database and defining my entities. Using raw SQL elsewhere is a viable option, even if less pleasant. I am actually considering whether I could simply either create views or stored procedures to query what is needed and present it in a relational format, but this also has its drawbacks.

In terms of your query about when someone else figured it out, [here is a link](https://github.com/dotnet/efcore/issues/8527#issuecomment-423489599) to where it was previously discussed back in 2018. He later submitted a PR with the proposed code, but it coincided with some architectural changes and the contributor did not resubmit for some reason. You actually replied to the thread in 2024 it looks like.

**Comment by roji** [comment]
> I personally would be more than happy with even the ability to simply create the tables with ASNODE or ASEDGE that I can define in my code.

I understand, but I don't think it makes sense for EF to only support table creation with SQL Server graph support, and nothing else. I get why it can simplify your specific scenario, but using EF exclusively to create the tables - and then querying with raw SQL - doesn't really make sense to me as a feature. If we only do this, we also run the risk of mis-designing the APIs, since the full feature (including querying and updating) may require more modeling/metadata than the base ASNODE/ASEDGE support would.

This also falls under the generate bucket of "please add create table feature X": we get lots of these for various (more or less obscure) DDL features of various databases. It's unfortunately unfeasible for EF to cover the entirety of what all relational databases allow in CREATE TABLE; these scenarios are already possible to handle by utilizing SQL in your migrations - alongside other migration operations which EF does support. So it generally comes down to people wanting EF to do 100% of the work for them via nice APIs, which unfortunately is simply too much - it really is OK (and expected) to drop down to a manual CREATE TABLE DDL from time to time.

> In terms of your query about when someone else figured it out, https://github.com/dotnet/efcore/issues/8527#issuecomment-423489599 to where it was previously discussed back in 2018.

Thanks for linking to that! The querying scenario shown in that comment is interesting, but feels like it deals with a very small subset of what LINQ querying would involve (I sense various minefields ahead, as I wrote above). And even for that simple scenario, expressing a graph query in terms of a lot of `from` queries and then a `where` to correlate between them and express the MATCH operator seems, well, not ideal - the LINQ looks far more verbose (and less understandable) then the actual graph query it translates into; always a bad sign. We really need to be careful not to try to shove a square peg into a round hole here, and try to force everything to look like LINQ.
